web3/utils/contracts.py

```diff
@@ -164,24 +172,32 @@ def prepare_transaction(abi,
         fn_name,
         fn_args,
         fn_kwargs,
+        fn_is_fallback
     )
     return prepared_transaction
 
 
-def encode_transaction_data(abi, web3, fn_name, args=None, kwargs=None):
+def encode_transaction_data(abi, web3, fn_name, args=None, kwargs=None, is_fallback_fn=False):
     fn_abi, fn_selector, fn_arguments = get_function_info(
-        abi, fn_name, args, kwargs,
+        abi, fn_name, args, kwargs, is_fallback_fn
     )
     return add_0x_prefix(encode_abi(web3, fn_abi, fn_arguments, fn_selector))
 
 
-def get_function_info(abi, fn_name, args=None, kwargs=None):
+def get_function_info(abi, fn_name, args=None, kwargs=None, is_fallback_fn=False):
+    if is_fallback_fn and fn_name is None:
+        raise FunctionNotSpecified("Have to specify either function name or use a fallback function.")
+
     if args is None:
         args = tuple()
     if kwargs is None:
         kwargs = {}
 
-    fn_abi = find_matching_fn_abi(abi, fn_name, args, kwargs)
+    if is_fallback_fn:
+        fn_abi = get_fallback_func_abi(abi)
+    else:
+        fn_abi = find_matching_fn_abi(abi, fn_name, args, kwargs)
+
     fn_selector = encode_hex(function_abi_to_4byte_selector(fn_abi))
```

Comment:
The fallback function has no function selector and I don't think `function_abi_to_4byte_selector` knows how to handle this correctly.  The function selector should be set to `b''` if we're interacting with the fallback function.

Reply:
I don't know how sending data to fallback works. If there is data there, isn't there some risk that the first four bytes of the data accidentally matches a function selector that does exist, and then you're not interacting with the fallback when you think you should be?

Reply:
I won't be able to investigate that right now, so I guess if we don't have a clear answer at hand: the short-term fix is to disallow data of any kind to the fallback. We can revisit later.

Reply:
You can send data to the fallback.  If the first 4 bytes of the data matches a function selector the contract will instead call that function.  I think we should do one of the following.

- do a check on whether the first four data bytes matches any known function selectors from the ABI and throw an error.
- do the same check but allow for an extra parameter to the function to explicitly allow the call to go through anyways (this one is my preference)

Reply:
By the way, my reasoning for wanting to support sending data that does in-fact match an existing function selector is the following.  It's a minority use case but it's real and shouldn't be too hard to support.

- I want to call an existing function (like the ERC20 `transfer`)
- I want to call it with a custom data packet (there was a recent bug some exchanges exposed with them not padding data correctly, a security researcher would want to be able to use this feature to pen-test).

Admittedly, they can do the same thing with simply `web3.eth.sendTransaction` but I think supporting it through the contract interface would be nice.

We **could** also just expose a new api `contract.sendTransaction` which just auto-fills the `to` address and allows you to specify everything else.

Reply:
Sorry for the late reply, I will add the implementation based on your second suggestion.

Reply:
Yes this seems like the best solution:

> do the same check but allow for an extra parameter to the function to explicitly allow the call to go through anyways (this one is my preference)

@zheli that is the one you chose, right?

Reply:
@carver yep.

Reply:
@pipermerriam have some questions regarding to the implementation. Where should I put the check for the first four data bytes? 
I thought I have to add it after we get the result from `encode_abi(web3, fn_abi, fn_arguments, fn_selector`, but then I realize the result will always be `0x` since fallback function doesn't have arguments, and fn_selector is always empty (`b''`). Could you explain a bit more? Thanks!